<h1><center>Laboratorio 3: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Primavera 2024</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Eduardo Moya
- Ayudantes: Nicolás Ojeda, Melanie Peña, Valentina Rojas

### Equipo: **SUPER IMPORTANTE - notebooks sin nombre no serán revisados**

- Nombre de alumno 1: Maria Jesus Espinoza
- Nombre de alumno 2: Sofia Lazcano


### **Link de repositorio de GitHub:** [Repositorio](https://github.com/jesuow/LabPro)

## Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 6 días de plazo con descuento de 1 punto por día. Entregas Jueves a las 23:59.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquier material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas` para manejar datos tabulares.
- Aprender a utilizar la documentación de pandas (y en general, de cualquier librería de python) como también a buscar soluciones en la web.

> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

### Librerias utiles para el Lab 😸

In [333]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para plotear
%pip install --upgrade plotly
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler

15878.41s - pydevd: Sending message related to process being replaced timed-out after 5 seconds


Note: you may need to restart the kernel to use updated packages.


# 1. Rendimiento en Estudiantes 📚

Para este laboratorio deberán trabajar con el dataset `students_grades.json` el cual contiene información sobre el rendimiento de alumnos en distintos ramos.
El objetivo principal será el de realizar un trabajo de limpieza de prepración y exploración de datos utilizando exclusivamente `DataFrames` y sus métodos asociados.




In [334]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = 'Dirección donde tiene los archivos en el Drive'
except:
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


## 1. Carga e Inspección de Datos [1.0 Puntos]


<p align="center">
  <img src="https://i.imgur.com/44knnIf.png" width="500" height="500">
</p>

*Don Caguayo*, rector de la Universidad de la Cachaña, les comenta que actualmente la Universidad no tiene una subdirección de análisis de datos, pero que tiene un gran interés en formar una. Es por esto, que se contacta con ud. y su equipo para iniciar de forma exploratoria diversas tareas de análisis y consolidación de datos respectivos al rendimiento de sus alumnos.

Previo a su primera reunión con el rector, el rector le envía una copia de la base de datos, la cuál a usted y su equipo le encantaría examinar. Para esto, la primera tarea que debe realizar es abrir y cargar el dataset.


> **Notar que es sumamente importante el formato que posee cada columna** para su manipulación y que posiblemente ustedes deberán especificar algunos tipos de datos a mano. Ver la documentación del método `read_json` para mayor información.

A continuación se muestra un ejemplo de los dos primeros registros del dataset:

```json
[
  {
    "names": "Rita Courtney",
    "gender": "female",
    "race/ethnicity": "group B",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 37,
    "reading score": 46,
    "writing score": 46
  },
  {
    "names": "Madeline Fuller",
    "gender": "female",
    "race/ethnicity": "group C",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 124,
    "reading score": 142,
    "writing score": 142
  }
]
```





In [335]:
df_grades = pd.read_json("students_grades.json")
numeric_columns = ['math score', 'reading score', 'writing score']
df_grades[numeric_columns] = df_grades[numeric_columns].apply(pd.to_numeric, errors='coerce')
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1      Madeline Fuller  female        group C            some high school   
2        Daniel Rester    male        group E          associate's degree   
3     Charles Linstrom    male        group A           bachelor's degree   
4       Geraldine Hale  female        group A          associate's degree   
...                ...     ...            ...                         ...   
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
1         standard                    none       124.0          142.0   
2         standard                    none       156.0          196.0   
3         standard               completed        80.0           78.0   
4         standard                    none       131.0          166.0   
...            ...                     ...         ...            ...   
1395      standard                    none       193.0          142.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1398  free/reduced                    none       193.0          196.0   
1399      standard               completed        68.0           80.0   

      writing score  
0              46.0  
1             142.0  
2             193.0  
3              81.0  
4             166.0  
...             ...  
1395          193.0  
1396           81.0  
1397           62.0  
1398          193.0  
1399           76.0  

[1400 rows x 9 columns]

In [336]:
print("Tipos de datos iniciales:")
print(df_grades.dtypes)



Tipos de datos iniciales:
names                           object
gender                          object
race/ethnicity                  object
parental level of education     object
lunch                           object
test preparation course         object
math score                     float64
reading score                  float64
writing score                  float64
dtype: object



<p align="center">
  <img src="https://i.giphy.com/media/4FQMuOKR6zQRO/giphy.webp">
</p>


Posteriormente, el rector les comenta que la sección de TI posee una base de datos con más de **1000** filas, en donde cada fila representa un alumno matriculado. Sin embargo, cree fuertemente que existen errores en esta: solo tiene registro de **1000** matriculas. Por este motivo, especula que se han realizado malas prácticas al registrar dichos datos.

En consecuencia, el rector les solicita a ustedes como tarea inicial que exploren la base de datos y encuentre posibles falencias dentro de sus registros. Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**. Cualquier registro que posea un valor diferente a esta escala, **debería ser eliminado de la base de datos**.


> **Instrucciones**

**Cree una función** que tome un `DataFrame` y que le permita revisar si las diferentes variables que lo componen son coherentes y si es posible observar anomalías en estas.

**To-Do**
1. [ ] Imprimir un mensaje con las dimensiones del Dataframe.
2. [ ] Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame.
3. [ ] Mostrar los primeros 5, los últimos 5 elementos del DataFrame y realice un muestreo de 5 elementos de forma aleatoria.
4. [ ] Mostrar una descripción rápida de las variables numéricas del DataFrame (utilice un método visto en clases).
5. [ ] Imprimir un mensaje con la cantidad de valores nulos (investigue el método `isna()` y `sum()` de los DataFrames).
6. [ ] Imprimir un mensaje que detalle la cantidad de valores unicos (investigue `nunique()`) .
7. [ ] Imprimir un mensaje con el total de filas que contengan nombres duplicados (investigue `duplicated()` y su parámetro `subset`).
8. [ ] Mostrar las filas que contengan nombres sean duplicados (investigue `duplicated()` y su parámetro `subset`).

Finalmente, responda las siguientes preguntas (comente esto en la celda ubicada debajo de **justificación**.):

9. [ ] ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?


10. [ ] Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?

> **Nota:** Para imprmir utilice `print(...)`. Para mostrar `DataFrames`, utilice `display(...)` ya que este permite mostrar estos elementos como tablas.

> **Nota 2:** Utilice la documentación de [pandas](https://pandas.pydata.org/docs/index.html) como también StackOverflow para resolver cualquier problema o inconveniente.

> **Nota 3:** Si no ha trabajado con boxplot antes, el siguiente link puede serle de mucha ayuda: [boxplot](https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c)

**Respuesta:**

In [337]:

def exploratory_data_analysis(dataframe):
    print(f"Dimensiones del DataFrame: {dataframe.shape}")
    print(f"Columnas del DataFrame: {dataframe.columns}")
    print(f"Primeros 5 elementos: {dataframe.head(5)}")
    print(f"Ultimos 5 elementos: {dataframe.tail(5)}")
    print(f"Muestreo de 5 elementos: {dataframe.sample(5, random_state=10)}")
    print(f"Descripción rápida: {dataframe.describe().round(2)}")
    print(f"Cantidad de valores nulos: {dataframe.isna().sum()}")
    print(f"Cantidad de valores unicos: {dataframe.nunique()}")
    print(f"Total de filas con nombres duplicados: {dataframe.duplicated(subset=['names'], keep=False).sum()}")
    print("Filas con nombres duplicados:")
    display(dataframe[dataframe.duplicated(subset=['names'], keep=False)])


    pass


In [338]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    df_grades.melt(
        value_vars=["math score", "reading score", "writing score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

In [339]:
exploratory_data_analysis(df_grades)

Dimensiones del DataFrame: (1400, 9)
Columnas del DataFrame: Index(['names', 'gender', 'race/ethnicity', 'parental level of education',
       'lunch', 'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')
Primeros 5 elementos:               names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
1  standard                    none       124.0          142.0          142.0  
2  standard                    none 

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
8     Jonathan Clingan    male        group D                some college   
10       Nancy Bassett  female        group C                 high school   
...                ...     ...            ...                         ...   
1385    James Crawford    male        group D             master's degree   
1390   Gloria Bradford  female        group B                 high school   
1393       Adam Romero    male        group C                 high school   
1395  Alexander Dillon    male        group D          associate's degree   
1398      Joseph Mccoy    male        group D                some college   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
5         standard                    none        73.0           66.0   
6         standard               completed        81.0           75.0   
8         standard               completed        71.0           61.0   
10        standard                    none       124.0          196.0   
...            ...                     ...         ...            ...   
1385      standard                    none       193.0          147.0   
1390      standard                    none       124.0           64.0   
1393      standard               completed       124.0          193.0   
1395      standard                    none       193.0          142.0   
1398  free/reduced                    none       193.0          196.0   

      writing score  
0              46.0  
5              63.0  
6              76.0  
8              69.0  
10            193.0  
...             ...  
1385          131.0  
1390          193.0  
1393           49.0  
1395          193.0  
1398          193.0  

[493 rows x 9 columns]

**Justificación:**

```
9. ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?
En el DataFrame, se observan 1153 valores únicos en la columna names, lo que sugiere que hay 1153 alumnos registrados. Este número es inusual dado que se tiene un registro de solo 1000 matrículas, lo que podría indicar errores en los datos o problemas de contabilización incorrecta de algunos registros.

En cuanto a los valores nulos, se encontraron valores faltantes en las siguientes columnas de notas:

math score: 40 valores nulos
reading score: 40 valores nulos
writing score: 47 valores nulos

Además, se detectaron 493 filas con nombres duplicados, lo que indica que algunos alumnos tienen múltiples registros en el DataFrame. Esto podría deberse a errores en el registro o a entradas duplicadas que necesitan ser revisadas y corregidas.

10. Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?
El boxplot de las notas revela que, en promedio, los estudiantes obtienen mejores resultados en el test de lectura (reading score), seguido por el test de escritura (writing score) y, por último, el test de matemáticas (math score).

Sin embargo, un aspecto común en los tres atributos es que algunos estudiantes tienen puntuaciones máximas que exceden el rango esperado de 0 a 100, es decir, hay valores que superan 100 en las puntuaciones de los exámenes. Por ejemplo, en el test de matemáticas, el puntaje máximo registrado es 196, lo cual está fuera del rango normal. Esto sugiere que podría haber errores en los datos o que se deben revisar las entradas para asegurar que las puntuaciones estén dentro del rango adecuado.
```

## 2 Limpieza de Datos [1.0 puntos]




<p align="center">
  <img src="https://blogdatlas.files.wordpress.com/2020/07/datlas_mx_blog_limpieza_bds_data-scrubbing-service.gif">
</p>

En base a lo realizado anteriormente, realice una limpieza de los dataset. Asuma los supuestos señalados en la pregunta anterior para limpiar el `DataFrame` y **cree una función** que le permita limpiar los datos. Habiendo limpiado el `DataFrame`, compruebe nuevamente los datos con la función creada en la sección 1.1.

**To-Do:**
- [ ] Eliminar registros con valores nulos.
- [ ] Eliminar valores fuera del rango válido.
- [ ] Eliminar registros con nombres duplicados
- [ ] Comprobar que no existen anomalías en los datos a traves de la función de la sección 1.1.

**Hints**

- Puede ser útil los métodos ```.drop_duplicates()``` y ```.dropna()``` como también utilizar queries con máscaras booleanas.
- Cuidado con el orden en la limpieza de datos. ¿Que se debe eliminar primero, los nombres duplicados o los registros fuera del rango posible?


> Utilice la función creada en el punto 1 para verificar sus resultados.

**Respuesta:**

In [340]:
# Escriba su código aquí
def clean_data(dataframe):
    dataframe = dataframe.dropna()
    dataframe = dataframe[ dataframe[numeric_columns].apply(lambda x: x.between(0, 100).all(), axis=1) ]
    dataframe = dataframe.drop_duplicates(subset=['names'], keep='first')
    
    return dataframe
    pass

In [341]:
#Pruebe su código aquí
df_grades = clean_data(df_grades)
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
3         standard               completed        80.0           78.0   
5         standard                    none        73.0           66.0   
6         standard               completed        81.0           75.0   
7         standard               completed        94.0           86.0   
...            ...                     ...         ...            ...   
1392      standard                    none        69.0           75.0   
1394  free/reduced               completed        23.0           44.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1399      standard               completed        68.0           80.0   

      writing score  
0              46.0  
3              81.0  
5              63.0  
6              76.0  
7              87.0  
...             ...  
1392           71.0  
1394           36.0  
1396           81.0  
1397           62.0  
1399           76.0  

[1000 rows x 9 columns]

In [342]:
exploratory_data_analysis(df_grades)

Dimensiones del DataFrame: (1000, 9)
Columnas del DataFrame: Index(['names', 'gender', 'race/ethnicity', 'parental level of education',
       'lunch', 'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')
Primeros 5 elementos:               names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
3  standard               completed        80.0           78.0           81.0  
5  standard                    none 

Empty DataFrame
Columns: [names, gender, race/ethnicity, parental level of education, lunch, test preparation course, math score, reading score, writing score]
Index: []

**Justificación:**

```
Después de limpiar el DataFrame, ahora se observa que hay 1000 valores únicos en la columna names, lo que corresponde al número esperado de matrículas, eliminando la anomalía previa de 1153 valores únicos que sugería errores en los datos.

En cuanto a los valores nulos, se ha verificado que no hay valores nulos en ninguna de las columnas, lo que indica que todos los campos están completos.

Además, se han eliminado los registros con nombres duplicados, por lo que ahora no hay filas con nombres repetidos en el DataFrame. Esto sugiere que los registros de alumnos están correctamente contabilizados y no hay entradas duplicadas.


```

## 3. Trabajando con Datos [2.0 puntos]

<p align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif">
</p>

### 3.1 Transformación de notas a "Nota Chilena" [0.3 Puntos]

En una de las reuniones con Don Caguayo este les comenta su intención de pasar todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree una función que le permita pasar a la escala Chilena las notas del dataframe y entregue las notas redondeadas al segundo decimal.

**To-Do**

- [ ] Tranforme los valores a la escala chilena.

**Respuesta:**

In [343]:
# Escriba su código aquí
def nota_chilena(dataframe_in, columna):
      # Aplicar la fórmula para convertir las notas a la escala chilena
  dataframe_in[columna] = (dataframe_in[columna] * 0.06) + 1
    
    # Redondear los valores al segundo decimal
  dataframe_in[columna] = dataframe_in[columna].round(2)
    
  return dataframe_in
  
  pass

In [344]:
nota_chilena(df_grades, "math score")
nota_chilena(df_grades, "reading score")
nota_chilena(df_grades, "writing score")

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score  
0              3.76  
3              5.86  
5              4.78  
6              5.56  
7              6.22  
...             ...  
1392           5.26  
1394           3.16  
1396           5.86  
1397           4.72  
1399           5.56  

[1000 rows x 9 columns]

In [345]:
#Muestre sus resultados aquí
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score  
0              3.76  
3              5.86  
5              4.78  
6              5.56  
7              6.22  
...             ...  
1392           5.26  
1394           3.16  
1396           5.86  
1397           4.72  
1399           5.56  

[1000 rows x 9 columns]

### 3.2 Obtención de promedio de alumnos [0.3 Puntos]

Con todas las notas adjuntadas, la prestigiosa Universidad De la Cachaña necesita obtener el promedio general de cada uno de los alumnos. Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del `DataFrame` original. Cabe señalar que las notas deben ser redondeadas al segundo decimal.

**To-Do:**
- [ ] Crear función que nos permite obtener el promedio de cada uno de los alumnos y anexar esta información en el `DataFrame` original.


**Respuesta:**

In [346]:
#Escriba su código aquí
def alumnos_promedio(dataframe_in):
    dataframe_in['promedio'] = dataframe_in[['math score', 'reading score', 'writing score']].mean(axis=1).round(2)
    return dataframe_in



In [347]:
#Muestre sus resultados aquí
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    none        3.22           3.76           3.76   
3  standard               completed        5.80           5.68           5.86   
5  standard                    none        5.38           4.96           4.78   
6  standard               completed        5.86           5.50           5.56   
7  standard               completed        6.64           6.16           6.22   

   promedio  
0      3.58  
3      5.78  
5      5.04  
6      5.64  
7      6.34

### 3.3 Reprobados y Destacados [0.9 Puntos, 0.3 cada DataFrame]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año y los que han tenido desempeño de destacado. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena, y se considerará destacado si posee notas superiores a 5.8 en todas las asignaturas (o sea notas superiores a 5.8 en `math score`, `reading score` y `writing score`).

Cree una función que permita obtener tres Dataframes:
- El primero deberá poseer solamente los alumnos que han reprobado en la prestigiosa universidad.
- El segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.
- El tercero será un Data Frame con los alumnos destacados ordenados de mayor a menor.

**To-Do:**

- [ ] Obtener DataFrame con los alumnos que reprobaron el año.
- [ ] Obtener DataFrame similar al original, pero marcados con una "R" el atributo GPA de los alumnos reprobados.
- [ ] Obtener DataFrame con los alumnos destacados ordenados de mayor a menor.

**Respuesta:**

In [348]:
#Escriba su código aquí
def reprobados_destacados(dataframe_in):
    #primer data frame s010 a1umn0s cuy0 pr0medi0 es men0r a 4,0
    boolean_mask= dataframe_in.loc[:,"promedio"]< 4.0
    reprobados= dataframe_in.loc[boolean_mask]

     #Crear una copia del DataFrame
    df = dataframe_in.copy()
    #Convertir la columna pr0medi0 a tipo float para asegurarse de que las comparaciones funcionen
    df["promedio"] = pd.to_numeric(df["promedio"], errors='coerce')
    df.loc[df["promedio"] < 4.0, "promedio"] = 'R'

    #Data Frame con los alumnos destacados ordenados de mayor a menor.
    destacados = dataframe_in[
        (dataframe_in['math score'] > 5.8) &
        (dataframe_in['reading score'] > 5.8) &
        (dataframe_in['writing score'] > 5.8)
    ]
    destacados = destacados.sort_values(by='promedio', ascending=False)

    return reprobados, df, destacados


In [349]:
# Muestre sus resultados aquí
dataframe1, dataframe2, dataframe3 = reprobados_destacados(df_grades)
dataframe1

/tmp/ipykernel_375336/1892390004.py:11: FutureWarning:

Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'R' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.



names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
13        Mikki Amaya  female        group A                 high school   
26      Samuel Dougal    male        group C             master's degree   
55    Rodger Talavera    male        group B                 high school   
63     Morris Johnson    male        group A          associate's degree   
...               ...     ...            ...                         ...   
1371   Charles Harian    male        group C            some high school   
1373    Wendy Gunnell  female        group D            some high school   
1389      Diana Costa  female        group C                 high school   
1391      Amy Sieving  female        group C                 high school   
1394    Wanda Russell  female        group B                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
13    free/reduced               completed        3.04           3.88   
26    free/reduced               completed        3.76           3.52   
55    free/reduced                    none        3.16           2.74   
63    free/reduced               completed        3.40           4.30   
...            ...                     ...         ...            ...   
1371  free/reduced               completed        3.70           4.12   
1373  free/reduced                    none        2.62           3.04   
1389  free/reduced                    none        3.46           3.76   
1391      standard                    none        2.74           2.74   
1394  free/reduced               completed        2.38           3.64   

      writing score  promedio  
0              3.76      3.58  
13             3.46      3.46  
26             3.76      3.68  
55             2.62      2.84  
63             4.18      3.96  
...             ...       ...  
1371           3.94      3.92  
1373           2.92      2.86  
1389           3.58      3.60  
1391           2.80      2.76  
1394           3.16      3.06  

[103 rows x 10 columns]

In [350]:
dataframe2

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score promedio  
0              3.76        R  
3              5.86     5.78  
5              4.78     5.04  
6              5.56     5.64  
7              6.22     6.34  
...             ...      ...  
1392           5.26      5.3  
1394           3.16        R  
1396           5.86     5.42  
1397           4.72     4.54  
1399           5.56     5.48  

[1000 rows x 10 columns]

In [351]:
dataframe3

names  gender race/ethnicity parental level of education  \
677       Richard Pryor    male        group E           bachelor's degree   
384   Elizabeth Beasley  female        group E           bachelor's degree   
1352      Sandra Pompey  female        group E          associate's degree   
475           Emma Gray  female        group E           bachelor's degree   
1276     Angela Cressey  female        group D                some college   
...                 ...     ...            ...                         ...   
949      Rolando Widmer    male        group C                 high school   
300        Steven Kelly    male        group D                some college   
296   Lucille Stanphill  female        group C                 high school   
1307        Peter Perez    male        group D             master's degree   
96          Harry Berry    male        group B          associate's degree   

         lunch test preparation course  math score  reading score  \
677   standard               completed        7.00           7.00   
384   standard                    none        7.00           7.00   
1352  standard                    none        7.00           7.00   
475   standard               completed        6.94           7.00   
1276  standard                    none        6.88           7.00   
...        ...                     ...         ...            ...   
949   standard               completed        5.92           6.04   
300   standard                    none        5.86           5.92   
296   standard                    none        5.86           6.04   
1307  standard                    none        5.86           5.86   
96    standard               completed        5.86           5.92   

      writing score  promedio  
677            7.00      7.00  
384            7.00      7.00  
1352           7.00      7.00  
475            7.00      6.98  
1276           6.94      6.94  
...             ...       ...  
949            5.92      5.96  
300            6.04      5.94  
296            5.92      5.94  
1307           6.04      5.92  
96             5.92      5.90  

[110 rows x 10 columns]

In [352]:
dataframe1["promedio"].max()
#c0mpr0bam0s que e1 pr0medi0 sea men0r a 4,0

3.98

### 3.4 Quintiles de Rendimiento [0.5 Puntos]

El rector, intrigado por saber si existe una relación entre el nivel educacional de los padres y las notas de sus hijos les plantea estudiar dicha relación. Para esto, les solicita que desde el `DataFrame` que posee a todos los alumnos (`DataFrame` obtenido tras la limpieza), generen una nueva columna que clasifique en quintiles de rendimiento para sus alumnos: `['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']`. (investigar `qcut()`)

Hecho esto, grafique a través de un scatter plot el nivel educacional de los padres v/s el promedio de los alumnos, utilice los cuantiles generados anteriormente para mostrar en que cuantil se encuentra cada uno de los datos.

Realizado el grafico, responda la siguiente pregunta: **¿Es posible visualizar alguna tendencia en el gráfico?.**



**Respuesta:**

In [353]:
# Escriba su código para quintiles aquí
# Escriba su código para quintiles aquí
def quintiles(dataframe_in):
    dataframe_in['GPA cuantiles'] = pd.qcut(dataframe_in['promedio'], q=5, labels=['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto'])
"""
NOTA: La variable del dataframe que contenga los quintiles debe llamarse
'GPA cuantiles' para que el gráfico funcione.
"""

"\nNOTA: La variable del dataframe que contenga los quintiles debe llamarse\n'GPA cuantiles' para que el gráfico funcione.\n"

In [354]:
quintiles(df_grades)

In [356]:
# Código de gráfico del amigo ChatGPT
fig = px.scatter(
    df_grades.sort_values(by="promedio"),
    x="promedio",
    y="parental level of education",
    hover_data=["gender"],
    color="GPA cuantiles",
    title="GPA por Nivel de educación de los padres",
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig

```
Escriba aqui su respuesta
```

## 4. Operaciones con tablas [2.0 puntos]

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYWx5dTBkMGhhOTJmeGQzbDV5M21rNTN1NWVwMm1qbHd3cjh3Nmk4NSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/13HBDT4QSTpveU/giphy.gif">
</p>



### 4.1 Carga de datos y concatenar dataset [0.2 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente).

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: `students_grades_1.json` y `students_grades_2.json`.


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos.

**No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas**, recuerde que anteriormente ya se encargo de este tema.


Hint: investigue el método `concat`

**Respuesta:**

In [357]:
# Inserte su respuesta aquí
df_grades_1= pd.read_json('students_grades_1.json')
df_grades_2= pd.read_json('students_grades_2.json')

df_grades= pd.concat([df_grades_1, df_grades_2], ignore_index=True)

df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
995     Richard Young    male        group D                 high school   
996     Wanda Russell  female        group B                 high school   
997    Marina Zeigler  female        group C           bachelor's degree   
998     Laurie Carter  female        group B            some high school   
999      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
995      standard                    none        5.14           5.50   
996  free/reduced               completed        2.38           3.64   
997  free/reduced               completed        4.96           5.44   
998      standard               completed        4.24           4.66   
999      standard               completed        5.08           5.80   

     writing score   GPA GPA cuantiles  
0             3.76  3.58      Bajísimo  
1             5.86  5.78          Alto  
2             4.78  5.04         Medio  
3             5.56  5.64          Alto  
4             6.22  6.34      Muy alto  
..             ...   ...           ...  
995           5.26  5.30         Medio  
996           3.16  3.06      Bajísimo  
997           5.86  5.42          Alto  
998           4.72  4.54          Bajo  
999           5.56  5.48          Alto  

[1000 rows x 11 columns]

### 4.2 Análisis por Nivel Educacional Etnia de los Padres [0.8 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos.
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.


**Tabla de Resumen [0.5 Puntos]**


Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

Utilizar la tabla de resultados esperados como guía para desarrollar este punto.


**Análisis [0.3 Puntos]**

Finalmente, responda las siguientes preguntas sobre el gráfico de caja que viene implementado a continuación. En este se muestra el GPA con respecto al nivel educacional, donde la variable de color es la etnicidad.

- [ ]  ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?


**Respuesta:**

In [365]:
# Inserte su respuesta aquí

import pandas as pd

def generar_resumen(dataframe_in):
    # Calcular el promedio de las notas (GPA) y redondear a dos decimales
    dataframe_in['GPA'] = dataframe_in[['math score', 'reading score', 'writing score']].mean(axis=1).round(2)

    # Mapeo del nivel educativo de los padres
    mapeo = {
        'some high school': 'school',
        'some college': 'school',
        'high school': 'school',
        'bachelor\'s degree': 'college',
        'associate\'s degree': 'college',
        'master\'s degree': 'postgraduate'
    }
    dataframe_in['simple parental level of education'] = dataframe_in['parental level of education'].map(mapeo)

    # Calcular el promedio de las notas por 'race/ethnicity' y 'simple parental level of education'
    resumen = dataframe_in.groupby(['race/ethnicity', 'simple parental level of education'])[['GPA', 'math score', 'reading score', 'writing score']].mean().reset_index()

    # Contar la cantidad de alumnos por grupo
    conteo = dataframe_in.groupby(['race/ethnicity', 'simple parental level of education']).size().reset_index(name='count')

    # Unir el promedio y el conteo
    resumen = pd.merge(resumen, conteo, on=['race/ethnicity', 'simple parental level of education'])

    # Calcular el porcentaje de alumnos
    total_alumnos = resumen['count'].sum()
    resumen['percentage'] = (resumen['count'] / total_alumnos * 100).round(2).astype(str) + '%'

    # Redondear los valores de las notas a dos decimales
    resumen[['GPA', 'math score', 'reading score', 'writing score']] = resumen[['GPA', 'math score', 'reading score', 'writing score']].round(2)

    return resumen




# [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
#- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

#       some high school -> school
#       some college -> school
#       high school -> school
#      bachelor's degree -> college #       associate's degree -> college
#      master's degree -> postgraduate
#
 # Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


#- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
#- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
#- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

#Utilizar la tabla de resultados esperados como guía para desarrollar este punto.



In [366]:
generar_resumen(df_grades)

race/ethnicity simple parental level of education   GPA  math score  \
0         group A                            college  4.94        4.83   
1         group A                       postgraduate  4.80        4.46   
2         group A                             school  4.71        4.65   
3         group B                            college  5.14        5.03   
4         group B                       postgraduate  5.49        5.03   
5         group B                             school  4.79        4.69   
6         group C                            college  5.24        5.03   
7         group C                       postgraduate  5.14        5.02   
8         group C                             school  4.88        4.74   
9         group D                            college  5.17        5.06   
10        group D                       postgraduate  5.59        5.35   
11        group D                             school  5.08        4.99   
12        group E                            college  5.47        5.53   
13        group E                       postgraduate  5.74        5.48   
14        group E                             school  5.25        5.35   

    reading score  writing score  count percentage  
0            5.05           4.95     26       2.6%  
1            4.88           5.06      3       0.3%  
2            4.81           4.66     60       6.0%  
3            5.24           5.16     61       6.1%  
4            5.81           5.63      6       0.6%  
5            4.90           4.79    123      12.3%  
6            5.36           5.33    118      11.8%  
7            5.23           5.17     19       1.9%  
8            5.00           4.89    182      18.2%  
9            5.22           5.24     78       7.8%  
10           5.63           5.78     23       2.3%  
11           5.13           5.11    161      16.1%  
12           5.45           5.43     57       5.7%  
13           5.93           5.83      8       0.8%  
14           5.27           5.11     75       7.5%

In [370]:
fig = px.box(
    df_grades,
    x="simple parental level of education",
    y="GPA",
    color="race/ethnicity",
    hover_data=["GPA"],
)
fig.show()

    Análizar gráfica acá.

**Resultados Esperados**

Esta tabla ilustra la estructura del dataframe a conseguir. Los valores son referenciales, no necesariamente son los esperados.

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.83 |            5.05 |            4.95 |  4.94 |      26 | 2.6 %        |
|  1 |                  | postgraduate                         |         4.46 |            4.88 |            5.06 |  4.80 |       3 | 0.3 %        |
|  2 |                  | school                               |         4.65 |            4.81 |            4.66 |  4.71 |      60 | 6.0 %        |
|  3 | group B          | college                              |         5.03 |            5.24 |            5.16 |  5.14 |      61 | 6.1 %        |
|  4 |                  | postgraduate                         |         5.03 |            5.81 |            5.63 |  5.49 |       6 | 0.6 %        |
|  5 |                  | school                               |         4.69 |            4.90 |            4.79 |  4.79 |     123 | 12.3 %       |
|  6 | group C          | college                              |         5.03 |            5.36 |            5.33 |  5.24 |     118 | 11.8 %       |
|  7 |                  | postgraduate                         |         5.02 |            5.23 |            5.17 |  5.14 |      19 | 1.9 %        |
|  8 |                  | school                               |         4.74 |            5.00 |            4.89 |  4.88 |     182 | 18.2 %       |
|  9 | group D          | college                              |         5.06 |            5.22 |            5.24 |  5.17 |      78 | 7.8 %        |
| 10 |                  | postgraduate                         |         5.35 |            5.63 |            5.78 |  5.59 |      23 | 2.3 %        |
| 11 |                  | school                               |         4.99 |            5.13 |            5.11 |  5.08 |     161 | 16.1 %       |
| 12 | group E          | college                              |         5.53 |            5.45 |            5.43 |  5.47 |      57 | 5.7 %        |
| 13 |                  | postgraduate                         |         5.48 |            5.93 |            5.83 |  5.74 |       8 | 0.8 %        |
| 14 |                  | school                               |         5.35 |            5.27 |            5.11 |  5.25 |      75 | 7.5 %        |


### 4.3 Joins, joins y más joins [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias.
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene registros de alumnos que ya no forman parte de la universidad, por lo que no deben ser parte de este análisis.


El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`).


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados o **alumnos que no pertenecen a la universidad**).

**To-Do**


- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

**Respuesta:**

In [369]:
# Inserte su respuesta aquí
other_grades = pd.read_csv('other_grades.csv')

df_grades_outer = pd.merge(df_grades, other_grades, how='outer', on='names')
print("Número de filas con outer join:", df_grades_outer.shape[0])
print(df_grades_outer.head())


df_grades_left = pd.merge(df_grades, other_grades, how='left', on='names')
print("Número de filas con left join:", df_grades_left.shape[0])
print(df_grades_left.head())

df_grades_right = pd.merge(df_grades, other_grades, how='right', on='names')
print("Número de filas con right join:", df_grades_right.shape[0])
print(df_grades_right.head())

df_grades_inner = pd.merge(df_grades, other_grades, how='inner', on='names')
print("Número de filas con inner join:", df_grades_inner.shape[0])
print(df_grades_inner.head())

Número de filas con outer join: 1150
          names gender race/ethnicity parental level of education  \
0  Aaron Morgan    NaN            NaN                         NaN   
1  Aaron Oakden   male        group D                 high school   
2  Aaron Shelly   male        group D          associate's degree   
3  Abigail Hall    NaN            NaN                         NaN   
4  Adam Collins    NaN            NaN                         NaN   

          lunch test preparation course  math score  reading score  \
0           NaN                     NaN         NaN            NaN   
1  free/reduced                    none        4.96           5.44   
2      standard                    none        4.66           3.88   
3           NaN                     NaN         NaN            NaN   
4           NaN                     NaN         NaN            NaN   

   writing score   GPA GPA cuantiles simple parental level of education  \
0            NaN   NaN           NaN                


Outer Join: Combina todas las filas de ambos DataFrames. Si un alumno solo está en uno de los DataFrames, aparecerá la fila, pero las columnas del otro DataFrame estarán vacías para ese alumno. Resultado: 1150 filas, con valores nulos para alumnos que no coinciden en ambos datasets.

Left Join: Muestra todas las filas del DataFrame original (df_grades), añadiendo las columnas del DataFrame de las nuevas notas (other_grades). Si un alumno no está en other_grades, sus columnas de historia y ciencias quedarán vacías. Resultado: 1000 filas.

Right Join: Muestra todas las filas del DataFrame other_grades, agregando datos de df_grades solo para los alumnos que coinciden. Alumnos en other_grades que no están en df_grades aparecerán con valores nulos en las demás columnas. Resultado: 1150 filas.

Inner Join: Solo incluye los alumnos que están presentes en ambos DataFrames, asegurando que todos los valores están completos. Resultado: 1000 filas, sin valores nulos.

Si quiero incluir únicamente a los alumnos que están en ambos DataFrames, el inner join es la mejor opción. Esto garantiza que las notas de ambas asignaturas (historia y ciencias) estén disponibles, eliminando cualquier registro que no las tenga y evitando inconsistencias en el análisis.

Sin embargo, si prefiero conservar a todos los alumnos del DataFrame original y solo agregar las notas de las nuevas asignaturas donde estén disponibles, el left join sería la opción más adecuada. Dado que ambos métodos me devuelven 1000 filas, en este caso no hay diferencia significativa entre usar uno u otro, por lo que cualquiera de los dos sería válido.

# Conclusión


Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>